### PR TITLE
[workflows] Add post-commit job that periodically runs the clang static analyzer

### PR DESCRIPTION
**Pull request by tstellar**
This job will run once per day on the main branch, and for every commit on a release branch.  It currently only builds llvm, but could add more sub-projects in the future.

OpenSSF Best Practices recommends running a static analyzer on software before it is released: https://www.bestpractices.dev/en/criteria/0#0.static_analysis

### CODE REVIEW

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (boomanaiden154) [inline .github/workflows/ci-post-commit-analyzer.yml:38]: Can you use the https://apt.llvm.org so we can specify the exact version of this and use something more recent than whatever Ubuntu 22.04 uses (I think v14?)?

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (boomanaiden154) [inline .github/workflows/ci-post-commit-analyzer.yml:52]: What's the benefit here of doing a dynamic library build?

**Review by boomanaiden154 — COMMENTED** [review]
Thanks for doing this.

I guess we'll see from the PR run how long this takes and what the results are like. From what I've heard, the clang static analyzer produces false positives that can be hard to fix. I'm not sure how that anecdote generalizes to the LLVM code base though.

Also, is there a reason for using clang static analyzer over something more basic like clang tidy? I think they both would fit the OpenSSF definition of static analysis (although haven't looked into it). I guess CSA does do more in depth analysis and will probably find more things.

Even if it's noisy, just having it run once a day post commit so that people who are interested can look at the results seems like a decent idea.

**Comment by tstellar** [comment]
> Thanks for doing this.
> 
> I guess we'll see from the PR run how long this takes and what the results are like. From what I've heard, the clang static analyzer produces false positives that can be hard to fix. I'm not sure how that anecdote generalizes to the LLVM code base though.
> 
> Also, is there a reason for using clang static analyzer over something more basic like clang tidy? I think they both would fit the OpenSSF definition of static analysis (although haven't looked into it). I guess CSA does do more in depth analysis and will probably find more things.
> 
> Even if it's noisy, just having it run once a day post commit so that people who are interested can look at the results seems like a decent idea.

This is really only for informational purposes at this point.  I was going to watch it for a while to see which of the checks are helpful and which ones are prone to false positives.  I'm open to changing the set of checks and also using clang-tidy if someone has a good recommendation for which checks to run.

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:52]: I just did that because it makes the build faster and takes less disk space.  I didn't think dynamic vs static matters for the static analyzers.

**Comment by steakhal** [comment]
> > Even if it's noisy, just having it run once a day post commit so that people who are interested can look at the results seems like a decent idea.
> 
> This is really only for informational purposes at this point. I was going to watch it for a while to see which of the checks are helpful and which ones are prone to false positives. I'm open to changing the set of checks and also using clang-tidy if someone has a good recommendation for which checks to run.

Usually, clang-tidy is good for context insensitive rules, basically for the cases that can be easily detected by looking at the AST.
CSA is usually better if some state-machine describes a bug condition, which is attached to some value. For example, if we have a pointer to some resource, use it and then close/release it. And this is tracked across different statements and conditions.

W.r.t the FP TP rate of the checks, the default enabled checkers is usually a fairly conservative selection, which should produce low FP rate in practice. There are useful checkers and combinations that are not enabled by default, but considered to be fairly mature otherwise. At Sonar, we use CSA as our symbolic execution engine backend for our rules where it's relevant, I'll ask if we can publish what subset of checkers we use in production. ([SonarCloud](https://www.sonarsource.com/products/sonarcloud/) is free for open-source projects, [example](https://github.com/sonarsource-cfamily-examples/linux-cmake-gh-actions-sc)).
I can also see the benefit of having something from directly in-house, like scan-build.

Besides the scan-build, there is also [CodeChecker](https://github.com/Ericsson/codechecke) which serves as a frontend for static analyzer invocations. They already maintain checker ["profiles"](https://github.com/Ericsson/codechecker/blob/master/config/labels/analyzers/clangsa.json) that categorizes the checkers. /cc @NagyDonat representing them. He could probably propose some refined list for us about what checkers work for them in practice.

For now, I think even the most basic setup is better than having nothing. So I'm all for having this PR.

---

FYI I also considered contributing something like this, but I found no policies around introducing CI workflows, and I'm aware that even if this is an open-source and well-known project, at the end of the day we probably only have limited CI time and/or some companies pay for that time - like build bots.

BTW don't we already have some nightly for scan-build? https://llvm.org/reports/scan-build/
Do you plan to replace that?

**Comment by tstellar** [comment]
@steakhal Thanks for the information.  CodeChecker looks interesting, but setting up a server is a little too much work for me.  

> FYI I also considered contributing something like this, but I found no policies around introducing CI workflows, and I'm aware that even if this is an open-source and well-known project, at the end of the day we probably only have limited CI time and/or some companies pay for that time - like build bots.
> 

We don't have any official policies about adding new CI.  GitHub let's us use up-to 60 concurrent action runners, which it provides for free.  I think if the job is small or it doesn't need to run for every PR then there is probably capacity for t.

> BTW don't we already have some nightly for scan-build? https://llvm.org/reports/scan-build/ Do you plan to replace that?

I think that is generated as part of apt.llvm.org, but I don't plan on replacing it.

**Review by haoNoQ — COMMENTED** [review]
I'm very excited to see this!

A few years ago we've actually had some work done to improve quality on LLVM specifically, and even built an LLVM-specific checker (it finds `dyn_cast`s that aren't checked for null before use): https://discourse.llvm.org/t/gsoc-2019-apply-the-clang-static-analyzer-to-llvm-based-projects-final-report/52919/4

cc @Xazax-hun @RKSimon @sylvestre @csaba-dabis 

IIRC it was Sylvestre who was maintaining the old scan-build output bot, and Simon has been voluntarily fixing these warnings for a while.

**Comment by RKSimon** [comment]
We used to test on coverity as well: https://scan.coverity.com/projects/llvm - but that broke some time ago and has proven tricky to keep running

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:54]: We could.  I wasn't sure how to decide which ones to enable, so I just left them all out .  If you have a suggested list of projects, I can do some benchmarks to see how long the builds will take.

Comment (haoNoQ) [reply .github/workflows/ci-post-commit-analyzer.yml:54]: Long-term I don't see a strong need to exclude any project. As long as it's built with clang, the static analyzer may be able to figure something out.

For the initial evaluation it may be a good idea to stick to projects with coding style similar to LLVM, because they'll probably have similar true/false positive rates.

I think `clang` is definitely good to go, maybe let's start with that? (It'll also imply all the fun of having the static analyzer analyze the static analyzer.)

Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:54]: Adding clang puts us over the 6 hour timeout limit for the job.  Is there some way I can optimize the checker to take less time?  Can I use ccache?  We could also just let it run for 6 hours and then upload the results when it's done.  I guess it doesn't matter much if we can't build all of clang.

Comment (boomanaiden154) [reply .github/workflows/ci-post-commit-analyzer.yml:54]: I'm guessing we could get speed-ups with PGO+ThinLTO+BOLT assuming perf-training on a good workload. That would be quite a bit of work to get going though. Might be something to look into down the road if this ends up getting used. We could split jobs like we've started doing in other places, but that would also take significant effort to implement.

I don't think I like the idea of just running until the job times out and then uploading whatever is there. I think that might make some issues pop up/go away each invocation depending upon what files it gets, decreasing the signal to noise ratio.

Comment (haoNoQ) [reply .github/workflows/ci-post-commit-analyzer.yml:54]: Oh.

Hmm.

Yes, the static analyzer can be tweaked for performance. In fact the entire idea is that the static analyzer tries to "explore all possible execution paths" which fundamentally takes an infinite amount of time. So the only reason it terminates at all, is because it implements a bunch of "budgets", hard cutoffs to limit the exploration time. A limit on the total number of simulation steps across the entire graph of paths, a limit on the number of iterations spent in a loop, a limit on the number of nested function calls. (To be fair, it also tries to do its best within the budget, eg. it prioritizes the paths that lead to previously unexplored lines of code. So increasing those limits yields rapidly diminishing returns.)

There's the "user-friendly" "deep"/"shallow" analysis mode setting that tweaks a bunch of budgets at once to produce two "supported" configurations; the shallow mode is faster but less precise. Unfortunately I'm afraid that the shallow mode was historically fine-tuned for Objective-C so I wouldn't really recommend it for any other purposes.

I think the easiest/safest thing to do is to simply decrease `-analyzer-config max-nodes=225000` to a smaller value. That's the total number of simulation steps. It affects analysis time non-linearly because many "entry-point" functions are simulated in a finite amount of steps so they never reach the limit anyway. But either way it's the simplest option to tweak because when this budget runs out the analysis simply terminates. It will simply miss some deeper bugs but it shouldn't introduce a lot of new false positives. Every other option means that the static analyzer gets more time to explore a different part of the code so it becomes significantly more chaotic.

So, can you try
```
scan-build ... -analyzer-config max-nodes=150000
```
or
```
scan-build ... -analyzer-config max-nodes=75000
```
to see if you can fit into the 6-hour limit this way?

----

Also, well, it's sad that we have to sacrifice thoroughness in order to cover more ground. This means that we'll need to sacrifice most of the other sub-projects, unless they're so self-contained that we can set up a new job with just them.

I hope that eventually we'll either get faster machines, or figure out how to split analysis into several smaller jobs.

We should probably consider `-DLLVM_ENABLE_MODULES=1` as well. Fundamentally the static analyzer isn't affected by modules so it should be a pure win. Not by a lot: path exploration time would probably outweigh compilation time anyway. But it could buy us an hour or so.

There's also the part where the code is compiled normally during analysis, and the static analyzer is invoked in a separate process from normal compilation, so all the code actually gets compiled twice. At least up to the AST, and then it goes in different directions. The output produced by normal compilation isn't used for the analysis so this is fundamentally unnecessary and could in theory be cut. Except we still need to compile the binaries that we need for compiling the rest of the code, such as `TableGen`. So in case of LLVM this is hard to avoid.

----

> Can I use ccache?

That's an excellent question, I've somehow never even considered this. There's nothing fundamentally impossible about this: the static analyzer is fully deterministic so it'll produce the same output given the same translation unit and flags as input. But it'll probably be very annoying to marry ccache and scan-build, because scan-build comes with its own compiler wrapper. Also I'm not sure ccache would understand where our output is, because what scan-build passes to clang as `-o` is a directory.

So let me try and see if there's a straightforward solution, but if not, I probably won't have time to implement a proper solution.

It may also be a good idea to drop `scan-build` entirely and implement the whole thing with a tiny custom `CMAKE_CXX_COMPILER_LAUNCHER`. That's the whole point of `scan-build` anyway, it's just a hacky way to intercept compiler invocations from arbitrary build systems and append `--analyze` to them. A hand-crafted integration into your build system is almost always preferable if you have time for it.

Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:54]: > It may also be a good idea to drop `scan-build` entirely and implement the whole thing with a tiny custom `CMAKE_CXX_COMPILER_LAUNCHER`. That's the whole point of `scan-build` anyway, it's just a hacky way to intercept compiler invocations from arbitrary build systems and append `--analyze` to them. A hand-crafted integration into your build system is almost always preferable if you have time for it.

What would a custom `CMAKE_CXX_COMPILER_LAUNCHER` need to do?  Just call `clang` to compile the object and then `clang --analyze` to analyze it?

Comment (mstorsjo) [reply .github/workflows/ci-post-commit-analyzer.yml:54]: > It may also be a good idea to drop scan-build entirely and implement the whole thing with a tiny custom CMAKE_CXX_COMPILER_LAUNCHER. That's the whole point of scan-build anyway, it's just a hacky way to intercept compiler invocations from arbitrary build systems and append --analyze to them. A hand-crafted integration into your build system is almost always preferable if you have time for it.

If done this way, one would need to do some extra work to wrap up the analyze outputs and gather it all into a common report, like scan-build does, right? Outside of the context of this PR, I'd be interested in knowing how to do this (I'm not sure if there's any good docs that show examples on how to do this)?

Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:83]: `-analyzer-config max-nodes=75000` helped get the builds under the time out.  I've also switch to using a more optimized compiler and added ccache for the compiler that actually compiles the code, but not for the compiler that does the static analysis.

Comment (haoNoQ) [reply .github/workflows/ci-post-commit-analyzer.yml:83]: > What would a custom CMAKE_CXX_COMPILER_LAUNCHER need to do? Just call clang to compile the object and then clang --analyze to analyze it?

Yes, something like this:
```sh
#!/bin/sh
"$@"
"$@" --analyze --analyzer-output html -o $OUTPUT_DIR \
    -Xclang -analyzer-config -Xclang max-nodes=150000
```

> Also, would it be better to just compile the project normally and then use clang-tidy to run the analyzer using the compilation database?

This probably wouldn't be faster because you'll still need to regularly re-generate the compilation database. On a buildbot this implies "every time". Also clang-tidy doesn't know how to produce HTML output, and static analyzer reports are incredibly hard to read without a GUI that overlays them on top of your source code (even though technically they the CLI gives you all the same information).

But the upside would be that you'll get clang-tidy checks too, without the *third* recompilation. I wanted to properly "marry" them for a long time so that this wasn't a choice you had to make, but never got to finishing it :(

> If done this way, one would need to do some extra work to wrap up the analyze outputs and gather it all into a common report, like scan-build does, right?

Yes, you can bring `scan-build` back for this last part:
```sh
scan-build --generate-index-only $OUTPUT_DIR
```
Btw, this works because HTML reports are machine-readable due to nice magic comments:
```html
  147 <!-- BUGDESC Dereference of null pointer (loaded from variable 'x') -->
  148
  149 <!-- BUGTYPE Dereference of null pointer -->
  150
  151 <!-- BUGCATEGORY Logic error -->
  152
  153 <!-- BUGFILE /Users/artemdergachev/test/test.c -->
  154
  155 <!-- FILENAME test.c -->
  156
  157 <!-- FUNCTIONNAME foo -->
  158
  159 <!-- ISSUEHASHCONTENTOFLINEINCONTEXT 9c4d6241eb00bff2310e3e3391b46023 -->
  160
  161 <!-- BUGLINE 83 -->
  162
  163 <!-- BUGCOLUMN 10 -->
  164
  165 <!-- BUGPATHLENGTH 2 -->
  166
  167 <!-- BUGMETAEND -->
  168 <!-- REPORTHEADER -->
```
So `scan-build` doesn't need to talk to clang or observe analysis in real time in order to rebuild the index.

And if you like you can do a lot of custom post-processing on top of it, like filter out reports by some of these categories, or separate them into different directories to build a separate index file for each project.

Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:83]: > Yes, something like this:
> 
> ```shell
> #!/bin/sh
> "$@"
> "$@" --analyze --analyzer-output html -o "$OUTPUT_DIR" \
>      -Xclang -analyzer-config -Xclang max-nodes=150000
> ```

OK, let me try that.

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (boomanaiden154) [inline .github/workflows/ci-post-commit-analyzer.yml:53]: What's the specific binary that we need from this package? We should probably add a TODO to include it in `LLVM_DISTRIBUTION_COMPONENTS` for the CI container so we can have a hermetic toolchain.

Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:53]: We just need `scan-build` which is actually a perl script.

Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:53]: @boomanaiden154 #94543

Comment (haoNoQ) [reply .github/workflows/ci-post-commit-analyzer.yml:53]: You can use `scan-build` directly from the source directory if you like, there's no need to install it with `apt-get`.

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (haoNoQ) [inline .github/workflows/ci-post-commit-analyzer.yml:93]: Not sure but I suspect that you should remove the `/**` part now. Because now that you removed `scan-build` nobody is creating that subdirectory with today's date anymore. All the HTML files go straight into `analyzer-results` now.

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml:53]: @haoNoQ OK, I still think I'm going to put it on the container, because that gives us a more stable version to use.

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (haoNoQ) [inline .github/workflows/ci-post-commit-analyzer.yml:11]: `'clang/**'`?

File: .github/workflows/ci-post-commit-analyzer.yml
Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer.yml]: I'm running from the source for now until the container is updated.

**Review by haoNoQ — APPROVED** [review]
I don't see any other potential problems, everything makes sense to me and I'm very excited!

**Comment by haoNoQ** [comment]
Ok looks like something crashed? The log isn't very informative, could also be out-of-memory or something of that nature. can we pass some flags to make it generate reproducers?

Actually `scan-build` does this out of the box by re-running the crashed invocation and put the crash dump and the repros into the results folder, in a sub-folder named `failures`. We could do the same in our launcher script, and then `scan-build --generate-index-only` might even pick them up and show them on the `index.html` page!

It may also be a good idea to `|| true` the static analyzer invocation, so that static analyzer crashes don't block the rest of the build. Static analyzer invocations don't depend on each other, their output doesn't need to "link" to each other, so it's ok to ignore these crashes. Also it's not like we're using a freshly built clang we can fix to unblock ourselves.

Side note, if we used the clang-tidy workflow with compilation databases, it'd naturally allow us to use freshly built clang for analysis, because all analysis would be run strictly after the normal build is finished. (We could use that workflow even without `clang-tidy`, just grep all run-lines from the compilation database and add `--analyze` etc. to them.)

**Comment by tstellar** [comment]
> Ok looks like something crashed? The log isn't very informative, could also be out-of-memory or something of that nature. can we pass some flags to make it generate reproducers?

I think it's out of memory, I noticed this line:
```
2024-06-06T02:14:25.6030984Z Killed
```

**Comment by haoNoQ** [comment]
Hmm, in this case could this also be the source of performance problems? Like it's swapping or something? Should we try to decrease our `-j`?

If it's just one specific file that goes OOM every single time, then we can exclude that file from analysis by wrapping it in:
```
#ifndef __clang_analyzer__
// the file
#endif
```
But in this case I should take a look at this file. It could be that there's some really weird code that dodges all our limits. (Historically it's been very long initializer lists that we try to represent perfectly in symbolic memory and they count as 1 node for the purposes of `max-nodes`.)

And yeah we should `|| true` the analyzer command either way. No reason to block on this.

**Comment by haoNoQ** [comment]
The static analyzer is several times more memory-intensive than the compiler, just like it's several times slower than the compiler. But the size of the file is usually not an issue: we deallocate all our analysis-related data structures after every top-level function, going back down to the AST. So if one specific file consumes unusually large amount of memory, it's not because the file is big; it's because one specific function in this file triggers a bug. You can also display all top-level functions with `-Xclang -analyzer-display-progress`. The last one printed is the one in which we got stuck. But this could be too verbose for an everyday buildbot job.

Let me take a look if I can make reproducers work.

**Comment by tstellar** [comment]
@haoNoQ I think RISCVFoldMasks.cpp has a pathological case that uses a lot of memory.

I've made a few changes based on your suggestions.  I split build and analyze into completely separate steps now.  This should allow us to avoid building some unnecessary targets and the script I use to run the analyzer will keep running if one of the jobs fail.

File: .github/workflows/ci-post-commit-analyzer-run.py
Comment (haoNoQ) [inline .github/workflows/ci-post-commit-analyzer-run.py:20]:
```suggestion
    print("Error analyzing file: ", e)
```

Comment (tstellar) [reply .github/workflows/ci-post-commit-analyzer-run.py:20]: Doesn't the print function automatically put a space between arguments?

Comment (haoNoQ) [reply .github/workflows/ci-post-commit-analyzer-run.py:20]: Oh crap nvm I forgot about that lol.

**Review by haoNoQ — APPROVED** [review]
Aha yeah either way LGTM!

I'll take a look at that crash in a moment.

At some point we should just merge your python script directly into scan-build so that you didn't need to reinvent that wheel.

**Review by boomanaiden154 — APPROVED** [review]
LGTM.

It seems like we aren't utilizing all the time now (~180/360 minutes). I think that should probably be fine. We can always adjust limits later as we have a better feel for the job duration.

**Pull request merged**
